toolchain/base/value_store.h

```diff
@@ -0,0 +1,138 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_BASE_VALUE_STORE_H_
+#define CARBON_TOOLCHAIN_BASE_VALUE_STORE_H_
+
+#include "common/check.h"
+#include "common/ostream.h"
+#include "llvm/ADT/APInt.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringMap.h"
+#include "toolchain/base/index_base.h"
+
+namespace Carbon {
+
+// Corresponds to a StringRef.
+struct StringId : public IndexBase, public Printable<StringId> {
+  static const StringId Invalid;
+  using IndexBase::IndexBase;
+  auto Print(llvm::raw_ostream& out) const -> void {
+    out << "str";
+    IndexBase::Print(out);
+  }
+};
+constexpr StringId StringId::Invalid(StringId::InvalidIndex);
+
+// Corresponds to an integer value represented by an APInt.
+struct IntegerId : public IndexBase, public Printable<IntegerId> {
+  static const IntegerId Invalid;
+  using IndexBase::IndexBase;
+  auto Print(llvm::raw_ostream& out) const -> void {
+    out << "int";
+    IndexBase::Print(out);
+  }
+};
+constexpr IntegerId IntegerId::Invalid(IntegerId::InvalidIndex);
+
+// Corresponds to a RealValue.
+struct RealId : public IndexBase, public Printable<RealId> {
+  static const RealId Invalid;
+  using IndexBase::IndexBase;
+  auto Print(llvm::raw_ostream& out) const -> void {
+    out << "real";
+    IndexBase::Print(out);
+  }
+};
+constexpr RealId RealId::Invalid(RealId::InvalidIndex);
+
+// A simple wrapper for accumulating values, providing IDs to later retrieve the
+// value. This does not do deduplication.
+template <typename ValueT, typename IdT>
+class ValueStore {
+ public:
+  // Stores the value and returns an ID to reference it.
+  auto Add(ValueT value) -> IdT {
+    auto id = IdT(values_.size());
+    values_.push_back(std::move(value));
+    return id;
+  }
+
+  // Returns the value for an ID.
+  auto Get(IdT id) const -> const ValueT& {
+    CARBON_CHECK(id.is_valid());
+    return values_[id.index];
+  }
+
+ private:
+  llvm::SmallVector<ValueT> values_;
+};
+
+// Storage for StringRefs. The caller is responsible for ensuring storage is
+// allocated.
```

Comment:
Switched to DenseMap. This was my intent and I think I copied too much from SemIR without thinking it through. Note though, to use StringMap's storage, the values_.push_back would need to use the StringMap's copy of value, not value directly.

That said, do you think we _should_ use StringMap, and have this be responsible for storing strings? The downside is that for builtin identifiers, we'd be duplicating storage. That may be mitigated because identifiers are expected to be reused. The upside is it's easier for generated strings, both for literals and metaprogramming.

I'm on the fence about what's best here, I was defaulting to this because it felt more minimal, but could switch.